### PR TITLE
perf: add reverse indices for O(1) lookup in hot paths

### LINKED ISSUE #2316: perf: high futex contention and tokio scheduling overhead
## Problem

Profiling the gateway shows significant time spent in synchronization primitives and tokio task scheduling:

### Syscall Profile (10 second sample)
```
52.72%  futex (339,886 calls, 44,023 errors)
34.30%  restart_syscall
11.44%  epoll_pwait
```

The high futex error rate (13%) suggests lock contention between threads.

### CPU Profile
```
12.08%  tokio::runtime::context::scoped::Scoped<T>::set
 6.10%  tokio::runtime::scheduler::multi_thread::queue::Steal<T>::steal_into
 4.23%  tokio::sync::task::atomic_waker::AtomicWaker::register_by_ref
 3.36%  sharded_slab::pool::Pool<T,C>::get
```

## Analysis

1. **High context switching**: 12% CPU in `Scoped::set` suggests frequent task context switches
2. **Work stealing overhead**: 6% in `steal_into` suggests uneven task distribution across worker threads
3. **Wakeup overhead**: 4% in `AtomicWaker::register_by_ref` suggests many small tasks with frequent wake/sleep cycles

This pattern is consistent with many short-lived tasks being constantly scheduled and rescheduled, rather than fewer longer-running tasks.

## Potential Causes

- Tasks that poll frequently without making progress
- Overly fine-grained task spawning
- Lock contention in shared data structures (dashmap, channels, etc.)

## Potential Solutions

1. **Task batching**: Combine related small operations into larger tasks
2. **Reduce lock scope**: Review critical sections for lock duration
3. **Worker thread tuning**: Adjust tokio worker thread count based on actual parallelism needs
4. **Profile-guided optimization**: Identify specific hot paths causing contention

## Context

Gateway CPU usage is ~42% after 3 hours with minimal load. This was discovered while investigating intermittent ~280% CPU spikes.

[AI-assisted - Claude]

### PR DESCRIPTION
**Pull request by sanity**
## Problem

Issue #2316 identified high futex contention (52% of syscall time) and lock contention as major performance bottlenecks. Investigation found two hot paths with O(n) complexity that caused excessive lock acquisition:

1. **LiveTransactionTracker::remove_finished_transaction** - Called on every operation completion, this method was iterating through ALL peers to find which one contained the transaction to remove
2. **SeedingManager::prune_subscriber** - Called on peer disconnection, this used `alter_all` which iterates ALL contracts

## Solution

Add reverse indices to convert these O(n) operations to O(1):

### LiveTransactionTracker
- Added `peer_for_tx: DashMap<Transaction, SocketAddr>` reverse index
- `remove_finished_transaction` now does direct lookup instead of full iteration
- `prune_transactions_from_peer` also cleans up the reverse index

### SeedingManager  
- Added `contracts_per_subscriber: DashMap<SocketAddr, DashSet<ContractKey>>` reverse index
- New `prune_subscriber_by_addr(SocketAddr)` method for O(1) pruning
- All add/remove operations maintain both indices consistently

## Testing

- All existing tests pass (27 seeding tests, 7 live_tx tests, 135 ring module tests total)
- Added test for `prune_subscriber_by_addr` method

## Impact

These changes reduce lock contention by:
- Eliminating full-map iterations in frequently-called paths
- Reducing the number of concurrent lock acquisitions
- Using more targeted, single-key operations

Note: This is a partial improvement for #2316 - there may be additional contention sources that could benefit from similar optimizations.

Closes #2316

[AI-assisted - Claude]

### CODE REVIEW

**Comment by sanity** [comment]
@claude Please do a skeptical review of this PR

**Comment by iduartgomez** [comment]
> These changes reduce lock contention by:

This PR claims that, but has this been actually measured?

File: crates/core/src/ring/live_tx.rs [hunk 1]
```diff
@@ -7,26 +7,27 @@ use std::sync::Arc;
...
+    /// Reverse index: Transaction -> SocketAddr for O(1) lookup during removal.
+    /// Without this, remove_finished_transaction would need to iterate all peers.
+    peer_for_tx: Arc<DashMap<Transaction, SocketAddr>>,
 }
 
 impl LiveTransactionTracker {
     pub fn add_transaction(&self, peer_addr: SocketAddr, tx: Transaction) {
         self.tx_per_peer.entry(peer_addr).or_default().push(tx);
+        self.peer_for_tx.insert(tx, peer_addr);
     }
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/src/ring/live_tx.rs:29]: Fixed in 01eb65e4. Reordered to insert to peer_for_tx first. Added comment explaining the race condition prevention.

[AI-assisted - Claude]

File: crates/core/src/ring/live_tx.rs [hunk 2]
```diff
@@ -36,11 +37,17 @@ impl LiveTransactionTracker {
...
 
     pub(crate) fn prune_transactions_from_peer(&self, peer_addr: SocketAddr) {
-        self.tx_per_peer.remove(&peer_addr);
+        // Remove all transactions for this peer from the reverse index
+        if let Some((_, txs)) = self.tx_per_peer.remove(&peer_addr) {
+            for tx in txs {
+                self.peer_for_tx.remove(&tx);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/src/ring/live_tx.rs:54]: Added test `prune_transactions_from_peer_cleans_both_indices` in 01eb65e4 that verifies both `tx_per_peer` and `peer_for_tx` are cleaned when a peer is pruned.

[AI-assisted - Claude]

File: crates/core/src/ring/seeding.rs [hunk 3]
```diff
@@ -199,6 +237,10 @@ impl SeedingManager {
                 .position(|l| l.pub_key == peer.pub_key && l.socket_addr() == Some(peer.addr))
             {
                 subs.swap_remove(pos);
+                // Update reverse index
+                if let Some(contracts) = self.contracts_per_subscriber.get(&peer.addr) {
+                    contracts.remove(contract);
+                }
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/src/ring/seeding.rs]: Fixed in 01eb65e4 using `remove_if_mut` pattern to atomically check if set is empty and remove the entry. Also added test `test_reverse_index_cleanup_no_empty_entries` to verify no empty entries remain.

[AI-assisted - Claude]

File: crates/core/src/ring/seeding.rs [hunk 4]
```diff
@@ -65,7 +70,16 @@ impl SeedingManager {
...
+            // Remove from reverse index first
+            if let Some((_, subs)) = self.subscribers.remove(evicted_key) {
+                for sub in subs {
+                    if let Some(addr) = sub.socket_addr() {
+                        if let Some(contracts) = self.contracts_per_subscriber.get(&addr) {
+                            contracts.remove(evicted_key);
+                        }
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/src/ring/seeding.rs]: Added test `test_eviction_cleans_reverse_index` in 01eb65e4 that specifically verifies `contracts_per_subscriber` is cleaned when contracts are evicted. The test adds subscribers to contracts, triggers eviction via record_contract_access, and asserts the reverse index has no entries for evicted contracts.

[AI-assisted - Claude]

File: crates/core/src/ring/seeding.rs [hunk 5]
```diff
@@ -84,7 +98,16 @@ impl SeedingManager {
...
+            // Clean up subscribers and reverse index
+            if let Some((_, subs)) = self.subscribers.remove(key) {
+                for sub in subs {
+                    if let Some(addr) = sub.socket_addr() {
+                        if let Some(contracts) = self.contracts_per_subscriber.get(&addr) {
+                            contracts.remove(key);
+                        }
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/src/ring/seeding.rs:115]: Good point. The reverse index cleanup in `remove_seeded_contract` is now covered by `test_eviction_cleans_reverse_index` since eviction calls `remove_seeded_contract` internally. Could add a dedicated test for direct `remove_seeded_contract` calls if desired.

[AI-assisted - Claude]

File: crates/core/src/ring/seeding.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/src/ring/seeding.rs]: Fixed in 01eb65e4 using `remove_if_mut` pattern. The eviction cleanup now uses:
```rust
self.contracts_per_subscriber.remove_if_mut(&addr, |_, contracts| {
    contracts.remove(evicted_key);
    contracts.is_empty()
});
```
This atomically removes the entry when the set becomes empty.

[AI-assisted - Claude]

File: crates/core/src/ring/seeding.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/src/ring/seeding.rs:115]: Fixed in 01eb65e4 using the same `remove_if_mut` pattern for consistency across all removal paths (`record_contract_access` eviction, `remove_seeded_contract`, and `remove_subscriber`).

[AI-assisted - Claude]

File: crates/core/src/ring/seeding.rs [hunk 6]
```diff
@@ -175,20 +206,27 @@ impl SeedingManager {
...
+                            removed_addr = %addr,
+                            "seeding_manager: pruning subscriber by address"
+                        );
+                        subs.swap_remove(pos);
+                    }
+                }
             }
-            subs
-        });
+        }
```
[the remark this answers is not in the document]
Comment (sanity) [reply crates/core/src/ring/seeding.rs:238]: Good catch on the ordering. In 01eb65e4, `prune_subscriber_by_addr` now iterates through contracts and removes from the `subscribers` map using `remove_if_mut`, which also cleans up empty entries. The reverse index is removed atomically at the start since we're removing ALL contracts for this address anyway - if a contract is somehow missing from subscribers (shouldn't happen in normal operation), the reverse index entry would be stale regardless. Added `test_reverse_index_cleanup_no_empty_entries` to verify no orphans remain.

[AI-assisted - Claude]

**Comment by sanity** [comment]
@iduartgomez 

> This PR claims that, but has this been actually measured?

No, this hasn't been measured with profiling yet. The PR addresses algorithmic complexity issues identified in the code review:

**Before:**
- `remove_finished_transaction`: O(n) - iterated all peers to find a transaction
- `prune_subscriber`: O(n) - used `alter_all` to iterate all contracts

**After:**
- Both operations are O(1) using reverse index lookups

The hypothesis is that reducing iteration over concurrent data structures reduces lock hold times and thus contention, but you're right that this should be validated with profiling. The original issue #2316 showed 52% futex time - would be valuable to re-run those benchmarks with this change to confirm actual improvement.

Should I run benchmarks before this goes in?

[AI-assisted - Claude]

**Comment by iduartgomez** [comment]
> @iduartgomez
> 
> > This PR claims that, but has this been actually measured?
> 
> No, this hasn't been measured with profiling yet. The PR addresses algorithmic complexity issues identified in the code review:
> 
> **Before:**
> 
>     * `remove_finished_transaction`: O(n) - iterated all peers to find a transaction
> 
>     * `prune_subscriber`: O(n) - used `alter_all` to iterate all contracts
> 
> 
> **After:**
> 
>     * Both operations are O(1) using reverse index lookups
> 
> 
> The hypothesis is that reducing iteration over concurrent data structures reduces lock hold times and thus contention, but you're right that this should be validated with profiling. The original issue #2316 showed 52% futex time - would be valuable to re-run those benchmarks with this change to confirm actual improvement.
> 
> Should I run benchmarks before this goes in?
> 
> [AI-assisted - Claude]

This rtades space for compute, which may be perfectly fine, but N in this case is neglegible. In any case, is not gonna hurt.
